File: Python3/144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        # order: node -> left -> right
        def dfs(node, l):
            if not node:
                return 
            
            # add node.val to the list
            l.append(node.val)
            
            if node.left:
                dfs(node.left, l)
            if node.right:
                dfs(node.right, l)
                
        
        ans = []
        dfs(root, ans)
        return ans

File: Java/144. Binary Tree Preorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        
        LinkedList<TreeNode> stack = new LinkedList<>();
        LinkedList<Integer> answer = new LinkedList<>();
        
        if(root == null){
            return answer;
        }
        
        stack.add(root);
        
        while(!stack.isEmpty()){
            TreeNode node = stack.pollLast();
            if(node != null){
                answer.add(node.val);
                if(node.right != null){
                    stack.add(node.right);
                }
                if(node.left != null){
                    stack.add(node.left);
                }
            }
        }
        
        return answer;
    }
}